File: src/tree/PathSum.java
package tree;

/*
https://leetcode.cn/problems/path-sum/description/?envType=problem-list-v2&envId=tree
路径总和
给你二叉树的根节点 root 和一个表示目标和的整数 targetSum 。判断该树中是否存在 根节点到叶子节点 的路径，这条路径上所有节点值相加等于目标和 targetSum 。如果存在，返回 true ；否则，返回 false 。

叶子节点 是指没有子节点的节点。

输入：root = [5,4,8,11,null,13,4,7,2,null,null,null,1], targetSum = 22
输出：true
解释：等于目标和的根节点到叶节点路径如上图所示。
 */
public class PathSum {

    static boolean isExists(TreeNode node, int targetSum) {
        int remain = targetSum - node.val;
        TreeNode left = node.left;
        TreeNode right = node.right;
        if (left == null && right == null) {
            return remain == 0;
        }

        boolean leftFlag = false;
        boolean rightFlag = false;
        if (left != null) {
            leftFlag = isExists(left, remain);
        }
        if (right != null) {
            rightFlag = isExists(right, remain);
        }
        return leftFlag || rightFlag;
    }

    public boolean hasPathSum(TreeNode root, int targetSum) {
        if (root == null) {
            return false;
        }
        return isExists(root, targetSum);
    }


    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

}

File: rust/src/binary_tree/path_sum.rs
// Definition for a binary tree node.
use crate::binary_tree::tree_node::TreeNode;
use std::cell::RefCell;
use std::rc::Rc;

struct Solution;
impl Solution {
    pub fn has_path_sum(root: Option<Rc<RefCell<TreeNode>>>, target_sum: i32) -> bool {
        let Some(node) = root else {
            return false;
        };
        let node = node.borrow();
        let remain = target_sum - node.val;
        if node.left.is_none() && node.right.is_none() {
            return remain == 0;
        }
        return Self::has_path_sum(node.left.clone(), remain)
            || Self::has_path_sum(node.right.clone(), remain);
    }
    /*pub fn has_path_sum(root: Option<Rc<RefCell<TreeNode>>>, target_sum: i32) -> bool {
        if root.is_none() {
            return false;
        }
        let Some(node) = root;
        let n = node.borrow();
        let val = n.val;
        let left = n.left.clone();
        let right = n.right.clone();
        let remain = target_sum - val;
        if remain == 0 && left.is_none() && right.is_none() {
            return true;
        }
        let mut left_flag = false;
        let mut right_flag = false;
        if let Some(left) = left {
            left_flag = Self::has_path_sum(left.borrow().left.clone(), remain);
        }
        if let Some(right) = right {
            right_flag = Self::has_path_sum(right.borrow().right.clone(), remain);
        }
        return left_flag || right_flag;
    }*/
}
#[cfg(test)]
mod tests {
    use crate::binary_tree::path_sum::{Solution, TreeNode};
    use std::cell::RefCell;
    use std::rc::Rc;

    #[test]
    fn binary_tree_path_sum_test() {
        let left = Some(Rc::new(RefCell::new(TreeNode::new(2))));
        let root = Some(Rc::new(RefCell::new(TreeNode {
            val: 1,
            left: left,
            right: None,
        })));
        let sum = Solution::has_path_sum(root, 3);
        assert_eq!(true, sum);
    }
}
